### DISCUSSION TITLE
How can we help?

### DISCUSSION THREAD

**Discussion by akessner** [fmasa/wfrp-master#389] [Ideas]
Hi, there are two/three features I'd really like to see added to this app.
1. import npcs from modules (personally I want to add npcs from if looks could kill, The enemy within, and night of blood )
2. Random stat generation for players and npcs
3. Sharing/importing npcs from the community

I recognize these are probably not the easiest things to add to the app, so my question is how can we help do that?

**Comment by fmasa** [comment]
Hey @akessner,
I would love to add most of these, but my time is quite limited and I can't really commit to a specific timeframe. But I think it's valuable to go through these - it's usually surprisingly fast to implement stuff where the problem space is known. If you feel like trying to implement some of these, I can provide pointers and context for the existing code.
 
> 1. import npcs from modules (personally I want to add npcs from if looks could kill, The enemy within, and night of blood )

I wonder if some streamlined way to create NPCs would solve the same issue (some one screen UI where user could easily input everything necessary) :thinking: Because the PDF imports are extremely time-consuming and this is a huge tasks with many edge cases :grimacing: 

Steps:

- **Understanding the stat block formats**
  There are different stat block formats thorough the books (with almost every new book, the Cubicle 7 changes alters some formatting of various things) and there are edge cases that will be impossible to support (e.g. Night of Blood defines some stat blocks for multiple NPCs and their differences are mentioned in the text paragraph). **The first step would be going through major books and grouping the stat blocks by these different formats** (so it's clear which formats are the most important to support). There is an useful index of such stat blocks in [this document](https://docs.google.com/document/d/e/2PACX-1vQPC6RFf2pfLQa-jLQrBqp7lqD3lEqtd8C-IChxCwV5HmtuAkeFrGu5hkbqnN_eh5JYxeA9OXhvnB4E/pub). The result should ideally be a a matrix that would show the different combinations as well.

- **Implementing the first Character import for some of the supported books**
  For these books a lot of existing parsing logic is already implemented - there are lexers and test suites and we will see the potential problems.
  
Some problems I am already aware of:

- Parsing the data
    - Various formats of specialisations should be handled - some books use `Melee (Basic) 60, Melee (Polearm 72)`, some use `Melee (Basic 60, Polearm 72)`
    - Inconsistent trappings formatting - `Bow and 7 arrows`, `Rope (12 yards)` - should be try to do something smart or just create misc trapping with the same name?
- Creating NPC Character from parsed data
    - The stat block might reference Skills, Talents, Trappings/... that might not exist in the Party Compendium - what should happen then?

> Random stat generation for players and npcs

This was proposed in https://github.com/fmasa/wfrp-master/issues/254. There are many features hiding behind this simple sentence :slightly_smiling_face: Can you explain how you would imagine this to work? :slightly_smiling_face: 

> Sharing/importing npcs from the community

Do you mean a generic import/export feature? I'm tracking it at https://github.com/fmasa/wfrp-master/issues/388

Or rather some in-app way for sharing the NPCs with other users?

**Reply by akessner** [reply]
The inconsistent formats for imports does sound troubling. 
Yes, perhaps just making it easier to fill in characteristics, talents, traits and skills, and having those be exportable to json, and then a way to import json will be the better way to go. 

As for random stats, I guess im picturing 2 features.
1, rolls 2d10 per characteristic in order then adds the racial bonuses. 
2. rolls 2d10 per characteristic, but doesn't add the racial bonus until you assign the number to a specific characteristic.
